### ISSUE TITLE
3.0.8 breaks PKITS test 4.1.5 (which requires DSA parameter inheritance)

### ISSUE THREAD

**Issue by brucestephens** [openssl/openssl#20233]
Indeed, just decoding the certificate fails:
```
openssl x509 -noout -text -in ValidDSAParameterInheritanceTest5EE.crt
```

The output includes
```
        Subject Public Key Info:
            Public Key Algorithm: dsaEncryption
            Unable to load Public Key
40477373937F0000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:../crypto/x509/x_pubkey.c:458:
40477373937F0000:error:03000072:digital envelope routines:X509_PUBKEY_get0:decode error:../crypto/x509/x_pubkey.c:458:
        X509v3 extensions:
```

(Version 1.1.1t does not seem affected.)

**Comment by paulidale** [comment]
#19663 would have caught this before release if it were done

**Comment by OttoHollmann** [comment]
If I understand this issue correctly, with above commit you were trying to disable keys with at least one NULL parameter. But according to RFC 3279 keys with all parameters equal to NULL are allowed.

```
  The id-dsa algorithm syntax includes optional domain parameters.
  These parameters are commonly referred to as p, q, and g.  When
  omitted, the parameters component MUST be omitted entirely.
```
So we should modify condition in function `ossl_d2i_DSA_PUBKEY` from
https://github.com/openssl/openssl/blob/cd870db16348d0d09cb05b7393cf9281509c7795/crypto/x509/x_pubkey.c#L763
`if (p == NULL || q == NULL || g == NULL) {`
to
`if ((p == NULL || q == NULL || g == NULL) && !(p == NULL && q == NULL && g == NULL)) { `
to explicitly allow keys with all parameters equal NULL?

Am I correct or it's more complicated?

**Comment by t8m** [comment]
Yes, it is much more complicated. You cannot do anything with such empty parameters key unless you set the parameters afterwards. And yeah, the only code that does this is the certificate chain validation code. Unfortunately it is also potential for issues as CVE-2023-0217 so I am not sure it is worth supporting this awkward DSA parameter inheritance thing but risk other CVEs because one uses parameterless DSA keys/certificates improperly.

**Comment by t8m** [comment]
So if there is a real use-case we might want to revert. Just having a test case in some artificial test suite IMO does not give that.

**Comment by paulidale** [comment]
PKITS are widely used and applied I think we should do something here.

That doesn't mean making that test pass necessarily.  Documenting that it fails would be sufficient.

**Comment by brucestephens** [comment]
> PKITS are widely used and applied I think we should do something here.
> 
> That doesn't mean making that test pass necessarily. Documenting that it fails would be sufficient.

Having read the explanation, that's what I was assuming you'd end up doing. Needs documenting, but I doubt anyone uses this feature of DSA parameter inheritance so most likely dropping it has no practical consequence.
